### DISCUSSION TITLE
How to use pg_trgm with imperative?

### DISCUSSION THREAD

**Discussion by jalvespinto** [sqlalchemy/sqlalchemy#7641] [Usage Questions] [answered]
There as some string columns on my db that I would like to search by text, but I am having some difficulties. For example, I can have a string column that I would like to search using pg_trgm similarity function, but I can get it to work. Any ideas? tks once more.
```
products = Table(
    'products', mapper_registry.metadata,
    Column('name', String(255),),
)

Index('ix_products_name', products.c.name, postgresql_ops={
      "name": "gin_trgm_ops"}, postgresql_using='gin')
```
I tried a few options, include the two bellow, but with no luck:
```
    def search_by_trgm(self, description: str) -> list[model.Product]:
        return self.session.query(model.Product).order_by(orm.products.c.name.op('<->')(description)).first()

    def search_by_trgm(self, description: str) -> list[model.Product]:
        return self.session.query(model.Product).order_by(func.similarity(orm.products.c.name,description).desc()).first()
```

**Comment by zzzeek** [comment, marked as answer]
if you need the pg_trgm prefix, you would call func.pg_trgm.similarity(Model.column, text).  that should work fine.   Other than that feel free to indicate what result you are getting.

**Reply by jalvespinto** [reply]
I think I might be using order_by wrong, because I get a `(psycopg2.errors.InvalidSchemaName) schema "pg_trgm" does not exist`
```
    def search_by_trgm(self, description: str) -> list[model.Product]:
        return self.session.query(model.Product).order_by(func.pg_trgm.similarity(orm.products.c.name, description).desc()).first()
```

**Reply by zzzeek** [reply]
OK go to your psql command line and get the SQL you want to work there first.

**Reply by jalvespinto** [reply]
something like this:
```
select name, similarity(name,'text here') as sim_score
from products
where similarity(name,'text here') > 0
order by sim_score desc
limit 10;
```

**Reply by jalvespinto** [reply]
I tried this way, but still cannot get it to work:
```
    def search_by_trgm(self, description: str) -> list[model.Product]:
        sim_score = func.pg_trgm.similarity(
            orm.products.c.name, description).label('sim_score')
        return self.session.query(model.Product, sim_score).filter(sim_score > 0).order_by(sim_score).limit(10)
```

**Reply by zzzeek** [reply]
> something like this:
> 
> ```
> select name, similarity(name,'text here') as sim_score
> from products
> where similarity(name,'text here') > 0
> order by sim_score desc
> limit 10;
> ```

this SQL is represented in the SQL expression language as:

```py

from sqlalchemy import select, desc, func

stmt = select(Product.name, func.similarity(Product.name, "text here").label("sim_score")).where(
   func.similarity(Product.name, "text here") > 0
).order_by(desc("sim_score")).limit(10)
```

**Reply by jalvespinto** [reply]
great. tks again for the help

**Reply by jalvespinto** [reply]
Hi @zzzeek. Sorry to bother on this again, but how can I create a index for this? I tried the snipt below, but I am getting a error:
```

Index(
    'ix_products_name',
    func.similarity(products.c.name).label('sim_score'), postgresql_ops={"sim_score": "gin_trgm_ops"}, postgresql_using='gin'
)
```
error:
```
E   sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedFunction) function similarity(character varying) does not exist
E   LINE 1: ...ATE INDEX ix_products_name ON products USING gin (similarity...
E                                                                                                              ^
E   HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
E   
E   [SQL: CREATE INDEX ix_products_name ON products USING gin (similarity(name) gin_trgm_ops)]
```

**Reply by zzzeek** [reply]
not sure.  I'm looking at https://www.postgresql.org/docs/9.1/pgtrgm.html and not seeing what CREATE INDEX syntax you're looking for.  You need to get the PostgreSQL syntax first, as I dont really deal with that end of it, then I can show you how to render it.

**Reply by jalvespinto** [reply]
ah, ok. I was looking here https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#operator-classes. I am not sure what the syntax would be, but I will try to find out. tks
